SchemaMigration/COSchemaMigrationDriver.m

```diff
@@ -42,28 +42,81 @@ static inline void addObjectForKey(NSMutableDictionary *dict, id object, NSStrin
 	[value addObject: object];
 }
 
+/**
+ * Returns some info on the metamodel state that the item entity/package/version
+ * comes from, namely the package/version pairs for the entity and all of its
+ * superclasses.
+ */
+- (NSDictionary *) versionsByDomainForItem: (COItem *)item
+{
+	NSDictionary *versionsByDomainByEntityTuple = [COSchemaMigration versionsByDomainByEntityTuple];
+	NSArray *key = @[item.packageName, @(item.entityVersion), item.entityName];
+	return versionsByDomainByEntityTuple[key];
+}
+
+
+
 - (NSSet *)domainsToMigrateForItem: (COItem *)item
 {
-	NSArray *domains = [item valueForAttribute: kCOObjectDomainsProperty];
 	ETEntityDescription *entity = [_modelDescriptionRepository descriptionForName: item.entityName];
 	BOOL isDeletedEntity = (entity == nil);
 	
 	/* For a deleted entity, the domain versions match between item and packages */
 	if (isDeletedEntity)
 	{
-		/* The first domain is the one owning the entity, the remaining domains
-		   own the parent entities. */
-		return S([domains firstObject]);
+		// FIXME: We hit this line when loading a CODictionary during the test suite,
+		// something must be wrong (but no tests fail).
+		return S(item.packageName);
+	}
+	
+	/* Early exit, the item has no version specified, or no package, (i.e.
```

Comment:
An important point is that we don't support package deletion, see -testPackageRenaming that explains why. 
So the case where the item has no package gives us compatibility with old CoreObject versions, e.g. the current  release, prior to the schema migration support. 

However in the long run, once we have done the next release (and migrated our existing stores), we should surely remove this and requires the package name to be never nil. We could add a TODO note.